src/implicit_function.jl

```diff
@@ -1,46 +1,134 @@
+"""
+struct Forward{handle_byproduct,F}
+    f::F
+    function Forward{handle_byproduct}(f::F) where {handle_byproduct,F}
+        return new{handle_byproduct,F}(f)
+    end
+end
 
-# Fields:
-- `forward::F`: callable of the form `x -> (ŷ(x),z(x))`.
-- `conditions::C`: callable of the form `(x,y,z) -> F(x,y,z)`
-- `linear_solver::L`: callable of the form `(A,b) -> u` such that `Au = b`, must be taken from Krylov.jl
+function Base.show(io::IO, forward::Forward{handle_byproduct}) where {handle_byproduct}
+    return print("Forward{$handle_byproduct}($(forward.f))")
+end
+
+"""
+    forward(x; kwargs...)
+
+Apply `forward.f` to `x`, returning a dummy byproduct `z(x)=0` if needed.
 """
-struct ImplicitFunction{F,C,L}
-    forward::F
-    conditions::C
-    linear_solver::L
+(forward::Forward{true})(x; kwargs...) = forward.f(x; kwargs...)
+(forward::Forward{false})(x; kwargs...) = (forward.f(x; kwargs...), 0)
+
+"""
+    Conditions{handle_byproduct,C}
+
+Callable wrapper for conditions `c::C`, which ensures that a byproduct `z` is always accepted in addition to `x` and `y`.
+
+The type parameter `handle_byproduct` is a boolean stating whether or not `c` natively accepts `z`.
+"""
+struct Conditions{handle_byproduct,C}
+    c::C
+    function Conditions{handle_byproduct}(c::C) where {handle_byproduct,C}
+        return new{handle_byproduct,C}(c)
+    end
 end
 
+function Base.show(
+    io::IO, conditions::Conditions{handle_byproduct}
+) where {handle_byproduct}
+    return print("Conditions{$handle_byproduct}($(conditions.c))")
+end
+
+"""
+    conditions(x, y, z; kwargs...)
+
+Apply `conditions.c` to `(x, y, z)`, discarding `z` beforehand if needed.
 """
-    ImplicitFunction(forward, conditions)
+(conditions::Conditions{true})(x, y, z; kwargs...) = conditions.c(x, y, z; kwargs...)
+(conditions::Conditions{false})(x, y, z; kwargs...) = conditions.c(x, y; kwargs...)
 
-Construct an `ImplicitFunction` with `Krylov.gmres` as the default linear solver.
 """
-function ImplicitFunction(forward, conditions)
-    return ImplicitFunction(forward, conditions, gmres)
+    ImplicitFunction{handle_byproduct,FF<:Forward,CC<:Conditions,LS}
+
+Differentiable wrapper for an implicit function defined by a forward mapping and a set of conditions.
+
+# Details
+
+- If `handle_byproduct=false`, the forward mapping is `x -> y(x)` and the conditions are `c(x,y(x)) = 0`.
+- If `handle_byproduct=true`, the forward mapping is `x -> (y(x),z(x))` and the conditions are `c(x,y(x),z(x)) = 0`. In this case, `z(x)` contains additional information generated by the forward mapping, but considered constant for differentiation purposes.
+
+Given `x ∈ ℝⁿ` and `y ∈ ℝᵈ`, we need as many conditions as output dimensions: `c(x,y,z) ∈ ℝᵈ`. We can then compute the Jacobian of `y(⋅)` using the implicit function theorem:
+```
+∂₂c(x,y(x),z(x)) * ∂y(x) = -∂₁c(x,y(x),z(x))
+```
+This requires solving a linear system `A * J = -B`, where `A ∈ ℝᵈˣᵈ`, `B ∈ ℝᵈˣⁿ` and `J ∈ ℝᵈˣⁿ`.
+
+# Fields
```

Comment:
I tried to keep coherent notations throughout:
- `forward / conditions` are our wrappers
- `f` / `c` are the callables provided by the user